src/caffe/layers/cudnn_conv_layer.cu

```diff
@@ -1,126 +1,182 @@
-    // Backward through cuDNN in parallel over groups and gradients.
-    for (int g = 0; g < this->group_; g++) {
-      // Gradient w.r.t. bias.
-      if (this->bias_term_ && this->param_propagate_down_[1]) {
-        CUDNN_CHECK(cudnnConvolutionBackwardBias(handle_[0*this->group_ + g],
-              cudnn::dataType<Dtype>::one,
-              top_descs_[i],  top_diff + top_offset_ * g,
-              cudnn::dataType<Dtype>::one,
-              bias_desc_, bias_diff + bias_offset_ * g));
-      }
+  template <typename Dtype>
+  void CuDNNConvolutionLayer<Dtype>::
+  Forward_gpu(const vector<Blob<Dtype>*>& bottom,
+              const vector<Blob<Dtype>*>& top) {
+    const Dtype* weight = this->blobs_[0]->gpu_data();
+    for (int i = 0; i < bottom.size(); ++i) {
+      const Dtype* bottom_data = bottom[i]->gpu_data();
+      Dtype* top_data = top[i]->mutable_gpu_data();
 
-      // Gradient w.r.t. weights.
-      if (this->param_propagate_down_[0]) {
-        const Dtype* bottom_data = bottom[i]->gpu_data();
-        CUDNN_CHECK(cudnnConvolutionBackwardFilter_v3(
-              handle_[1*this->group_ + g],
-              cudnn::dataType<Dtype>::one,
-              bottom_descs_[i], bottom_data + bottom_offset_ * g,
-              top_descs_[i],    top_diff + top_offset_ * g,
-              conv_descs_[i],
-              bwd_filter_algo_[i], workspace[1*this->group_ + g],
-              workspace_bwd_filter_sizes_[i],
-              cudnn::dataType<Dtype>::one,
-              filter_desc_, weight_diff + this->weight_offset_ * g));
+      // Test free space and force reshape if allocations have changed
+      size_t workspace_limit_bytes, total_memory;
+      gpu_memory::getInfo(&workspace_limit_bytes, &total_memory);
+      if (workspace_fwd_sizes_[i] > workspace_limit_bytes) {
+          this->Reshape(bottom, top);
       }
 
-      // Gradient w.r.t. bottom data.
-      if (propagate_down[i]) {
-        if (weight == NULL) {
-          weight = this->blobs_[0]->gpu_data();
+      // !!!! Not safe if group_ > 1 !!!!
+      workspace.reserve(workspace_fwd_sizes_[i]);
+
+      // Forward through cuDNN in parallel over groups.
+      for (int g = 0; g < this->group_; g++) {
+          // Filters.
+        CUDNN_CHECK(cudnnConvFwd(Caffe::cudnn_handle(),
+                                 cudnn::dataType<Dtype>::one,
+                                 bottom_descs_[i],
+                                 bottom_data + bottom_offset_ * g,
+                                 filter_desc_,
+                                 weight + this->weight_offset_ * g,
+                                 conv_descs_[i],
+                                 fwd_algo_[i],
+                                 workspace.data(),
+                                 workspace.size(),
+                                 cudnn::dataType<Dtype>::zero,
+                                 top_descs_[i],
+                                 top_data + top_offset_ * g));
```

Comment:
Again, could we maybe write these in logical groups instead of taking up thirteen lines?